src/java.base/share/classes/sun/security/pkcs12/MacData.java

```diff
@@ -72,28 +83,65 @@ class MacData {
         AlgorithmId digestAlgorithmId = AlgorithmId.parse(digestInfo[0]);
         this.digestAlgorithmName = digestAlgorithmId.getName();
         this.digestAlgorithmParams = digestAlgorithmId.getParameters();
+
         // Get the digest.
         this.digest = digestInfo[1].getOctetString();
 
-        // Get the salt.
-        this.macSalt = macData[1].getOctetString();
+        if (digestInfo[0].tag != DerValue.tag_Sequence) {
+            throw new IOException("algid parse error, not a sequence");
+        }
+        if (digestAlgorithmName.equals("PBMAC1")) {
+            PBEParameterSpec pbeSpec;
+
+            try {
+                pbeSpec =
+                        digestAlgorithmParams.getParameterSpec(
+                        PBEParameterSpec.class);
+            } catch (InvalidParameterSpecException ipse) {
+                throw new IOException(
+                        "Invalid PBE algorithm parameters");
+            }
+            iterations = pbeSpec.getIterationCount();
+            macSalt = pbeSpec.getSalt();
+            String ps = digestAlgorithmParams.toString();
+            kdfHmac = getKdfHmac(ps);
+            if (!(kdfHmac.equals("HmacSHA512") ||
+                    kdfHmac.equals("HmacSHA256"))) {
+                throw new IllegalArgumentException("unsupported PBMAC1 Hmac");
+            }
+
+            Hmac = kdfHmac;
+        }
+
+        // Get the old salt.
+        extraSalt = macData[1].getOctetString();
```

Comment:
Do we need the `extraSalt` and `extraIterations` variables? What if you just put an `else` block here (if not PBMAC1) and then fill in the `macSalt` and `iterations`.